Terraform Workstation as a Linux VM with MSI with optional Desktop

VaijanathB opened the pull request:
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

bmoore-msft commented on 201-vm-msi-linux-terraform/.gitignore:
```diff
@@ -0,0 +1 @@
+greg.parameters.json
```
remove this file

VaijanathB replied:
Done.

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+  },
+  "variables": {
+    "dnsLabelPrefix": "[concat('msi',uniquestring(resourceGroup().id,deployment().name))]",
+    "_artifactsLocation": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-vm-msi-linux-terraform",
```
See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc) for how to handle artifacts

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    }
+  },
+  "variables": {
+    "dnsLabelPrefix": "[concat('msi',uniquestring(resourceGroup().id,deployment().name))]",
```
remove deployment().name

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    "infraStorageAccountName": "[take(concat('storeinfra', uniquestring(resourceGroup().id), variables('dnsLabelPrefix')),24)]",
+    "stateStorageAccountName": "[take(concat('storestate', uniquestring(resourceGroup().id), variables('dnsLabelPrefix')),24)]",
+    "nicName": "[concat('nic',uniquestring(resourceGroup().id,deployment().name))]",
+    "networkSecurityGroupName": "[concat('nsg',uniquestring(resourceGroup().id,deployment().name))]",
```
remove deployment().name

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    "addressPrefix": "10.0.0.0/16",
+    "subnetName": "Subnet",
+    "subnetPrefix": "10.0.0.0/24",
+    "publicIPAddressName": "[concat('pip',uniquestring(resourceGroup().id,deployment().name))]",
```
remove deployment().name

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    "subnetName": "Subnet",
+    "subnetPrefix": "10.0.0.0/24",
+    "publicIPAddressName": "[concat('pip',uniquestring(resourceGroup().id,deployment().name))]",
+    "vmName": "[concat('vm',uniquestring(resourceGroup().id,deployment().name))]",
```
remove deployment().name

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    "subnetPrefix": "10.0.0.0/24",
+    "publicIPAddressName": "[concat('pip',uniquestring(resourceGroup().id,deployment().name))]",
+    "vmName": "[concat('vm',uniquestring(resourceGroup().id,deployment().name))]",
+    "virtualNetworkName": "[concat('vnet',uniquestring(resourceGroup().id,deployment().name))]",
```
remove deployment().name

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    "vmName": "[concat('vm',uniquestring(resourceGroup().id,deployment().name))]",
+    "virtualNetworkName": "[concat('vnet',uniquestring(resourceGroup().id,deployment().name))]",
+    "subnetRef": "[resourceId('Microsoft.Network/virtualNetworks/subnets', variables('virtualNetworkName'), variables('subnetName'))]",
+    "linuxConfigurationChosen": "[concat('linuxConfiguration', parameters('authenticationType'))]",
```
simplify this section around ssh or pw authn - see: https://github.com/Azure/azure-quickstart-templates/blob/master/201-vm-new-or-existing-conditions/azuredeploy.json

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+    },
+    {
+      "name": "creatingRBAC",
+      "type": "Microsoft.Resources/deployments",
```
why do we need a nested deployment for this?

bmoore-msft replied:
Those are probably just dated - let's move this to the main template and get rid of the nesting.

VaijanathB replied:
Removed the neseted Deployment.

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.json:
```diff
@@ -0,0 +1,320 @@
+          ]
+        },
+        "protectedSettings": {
+          "commandToExecute": "[concat('bash infra.sh && bash install.sh ', variables('installParm1'), variables('installParm2'), variables('installParm3'), variables('installParm4'), ' -k ', listKeys(resourceId('Microsoft.Storage/storageAccounts', variables('stateStorageAccountName')), providers('Microsoft.Storage', 'storageAccounts').apiVersions[0]).keys[0].value, ' -l ', reference(concat(resou ...
```
simplify you concat on the resourceId for the MSI - just use the resourceId function

bmoore-msft commented on 201-vm-msi-linux-terraform/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,15 @@
+    "adminUsername": {
+      "value": "GEN-UNIQUE"
+    },
+    "authenticationType": {
```
remove this since you have defaultValue

VaijanathB replied:
Done

bmoore-msft commented on 201-vm-msi-linux-terraform/nestedTemplates/setUpRbac.json:
```diff
@@ -0,0 +1,27 @@
+      }
+    },
+    "variables": {
+      "owner": "[concat('/subscriptions/', subscription().subscriptionId, '/providers/Microsoft.Authorization/roleDefinitions/', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')]"
```
any reason this can't be contributor instead of owner?

VaijanathB replied:
I believe MSI requires owner. The other two samples which do MSI have the same permission level.

bmoore-msft replied:
MSI doesn't need owner perms, contrib will work as long as you're not doing anything that requires owner (can't imagine what that would be) - so let's use least privilege here.

VaijanathB replied:
The owner was a variable name. Anyway I have changed the variable name  to contributor to avoid any confusion.

VaijanathB commented:
Hello Brian,

I have addressed all the comments in this PR. Some of them like nestedTemplates I am keeping them as it is as all other templates with MSI are following the same pattern.  Please let me know if you need more   information on these issues.  Thanks for taking time to review thoroughly. I really appreciate it. 

~VJ

VaijanathB commented:
Thanks for getting this done. 
~VJ
